### PR TITLE
Fix Cropbox with indices. Add new assertions for unit tests.

### LINKED ISSUE #2922: Wrong index in pcl::cropbox
Operating System and version: Ubuntu 16.04
Compiler: GCC
PCL Version: 1.7

I have some strange behavior in the Cropbox filter. If I use the filter without setting the indices via `Cropbox::setIndices(..)`, it works as expected. Once I set the indices, with only a subset of the whole pointcloud indices, the output is confusing. It seems that a lot of points are missing in the output.

I guess it has something to do with line 101 in the cropbox.hpp file. Shouldn't it be `(* indices_) [index]` instead of `index`? I think the selected point should be determined by the id  stored in  the `indices_` vector at this point.

And what about the lines 129 and 137? Is it the same there?

Can anybody confirm this?

### PR DESCRIPTION
**Pull request by Enderdead**
In order to cover the current crop box issue, I have added new pcl_tests functions. I'm not sure if you will agree with my adding so I'm waiting for your review about it.
The fix on CropBox class is pretty obvious just replace some index by (*indices_)[index]

Fixes #2922.

### CODE REVIEW

**Comment by taketwo** [comment]
Hi, we can add new test facilities, no problem. However, I would propose to implement them similarly to other assertions (e.g. equality of XYZ points). That is, have two parts:
1. Function that does actual test and returns `::testing::AssertionResult`.
2. Convenience macro that puts together the function and `ASSERT_PRED_FORMAT2` to provide a simpler interface to the user.

The function you have already, we only need to rename it to something like `VectorContainsAll` (note no "expect" because this function can be used both in "expect" and "assert" checks) and move to `internal` namespace.

The macro should be trivial to implement, just check the existing ones. Note that we need a pair of "expect" and "assert" macros so that the user can choose the one that is more appropriate for the task at hand.

**Comment by Enderdead** [comment]
Sorry I didn't see these macros on the bottom. 
I'm not really proud of the current log on internal::VectorContainsAll. 
In my initial commit, I wrote this :
```cpp
std::ostringstream vec_rep;
std::copy(v.cbegin(), v.cend()-1, std::ostream_iterator<V>(vec_rep, ", "));
vec_rep<<v.back();

std::ostringstream elements_rep;
std::copy(elements.cbegin(), elements.cend()-1, std::ostream_iterator<V>(elements_rep, ", "));
elements_rep << elements.back();

 return ::testing::AssertionFailure ()
       << "Vector       : "<< expr1<< std::endl
       << "contains     : " << vec_rep.str() << std::endl
       << "From         : " << expr2<< std::endl
       << "must contain : " << elements_rep.str() << std::endl;
```
This program will display log like this : 
```
/pcl/test/filters/test_clipper.cpp:356: Failure
Vector       : *removed_indices
contains     : 2, 3, 4, 6
From         : std::vector<int>({3, 5, 6, 8})
must contain : 3, 5, 6, 8
```
As you can see here, you have a duplicate target vector and if you get large vectors, it will be very hard to figure out the log. Do you have any ideas to improve that?

**Comment by taketwo** [comment]
I don't think we can do much about this. Perhaps do not print vector contents if its size is larger than a certain threshold. As for the duplication, it only happens if you supply "from" vector in-place. When you know that you are dealing with a large vector, create a temporary variable (e.g. `expected_result`) to hold it and then pass to the test macro.

**Comment by Enderdead** [comment]
Another way is to use the gmock lib (https://github.com/google/googletest/blob/master/googlemock/docs/cheat_sheet.md#container-matchers) . It provide many new predicates like `UnorderedElementsAre`.

**Comment by taketwo** [comment]
To be honest, I am reluctant to add a new dependency just to get a few predicates. If we really need them, I'd rather copy their implementation (_gmock_ is a BSD library, so should be ok). @SergioRAgostinho what's your stance?

**Comment by Enderdead** [comment]
I think you are right. While only test_clipper.cpp use vector asserts, it's pretty useless to add another dependency on the whole project.  Do I have to squash my new commit with the previous one?

File: common/include/pcl/pcl_tests.h [hunk 1]
```diff
@@ -219,6 +219,52 @@ namespace pcl
         return ::testing::AssertionSuccess ();
       }
 
+
+        template<typename V>
+        ::testing::AssertionResult VectorContainsAll(const char*  expr1, const char* expr2, std::vector<V> elements, std::vector<V> v) {
```
Comment (taketwo) [inline common/include/pcl/pcl_tests.h]: There is no reason to pass vectors by value (involves copy), let's do `const&` instead.

File: common/include/pcl/pcl_tests.h [hunk 2]
```diff
@@ -219,6 +219,52 @@ namespace pcl
...
+                    vec_rep<<v.back();
+
+                    std::ostringstream elements_rep;
+                    std::copy(elements.cbegin(), elements.cend()-1, std::ostream_iterator<V>(elements_rep, ", "));
+                    elements_rep << elements.back();
+
+                    return ::testing::AssertionFailure ()
+                            << "Vector       : "<< expr2 << std::endl
+                            << "contains     : " << vec_rep.str() << std::endl
+                            << "From         : " << expr1 << std::endl
```
Comment (taketwo) [inline common/include/pcl/pcl_tests.h]: I'm a bit confused how you intend to read this message. The first two lines I understand, "The actual vector ... contains ...". But what does "from" mean?

File: common/include/pcl/pcl_tests.h [hunk 3]
```diff
@@ -219,6 +219,52 @@ namespace pcl
...
+                            << "must contain : " << elements_rep.str() << std::endl;
+                }
+            }
+            return ::testing::AssertionSuccess ();
+        }
+
+        template<typename V>
+        ::testing::AssertionResult VectorDoesNotContain(const char*  expr1, const char*  expr2, std::vector<V> elements, std::vector<V> v) {
```
Comment (taketwo) [inline common/include/pcl/pcl_tests.h]: `const&` here as well.

File: test/filters/test_clipper.cpp [hunk 4]
```diff
@@ -39,6 +39,8 @@
  */
 
 #include <gtest/gtest.h>
+#include <gtest/gtest_pred_impl.h>
```
Comment (taketwo) [inline test/filters/test_clipper.cpp]: Not needed anymore?

File: test/filters/test_clipper.cpp [hunk 5]
```diff
@@ -287,7 +303,172 @@ TEST (CropBox, Filters)
...
+
+  // Indices
+  cropBoxFilter.filter (indices);
+
+  // Cloud
+  cropBoxFilter.filter (cloud_out);
+
+  // Should contain all
+  EXPECT_EQ (int (indices.size ()), 7);
+  EXPECT_VECTOR_CONTAINS_ALL (std::vector<int>({1, 2, 3, 5, 6, 7, 8}), indices);
```
Comment (taketwo) [inline test/filters/test_clipper.cpp]: Is it possible to write `{1, 2, 3, 5, 6, 7, 8}` here, does the compiler manage to deduct type? Would be cleaner.

Comment (Enderdead) [reply test/filters/test_clipper.cpp]: Unfortunately, MACRO doesn't figure out initializer-list and uses each element as a single parameter. I'm not really comfortable with variadic macro, I found some partial solutions but they don't cover all situations.

File: common/include/pcl/pcl_tests.h (hunk 2 above)
Comment (Enderdead) [reply common/include/pcl/pcl_tests.h]: Do you prefer this one ?
```suggestion
                             << "Actual       : " << expr2 << std::endl
                            << "contains     : " << vec_rep.str() << std::endl
                            << "Target set  : " << expr1 << std::endl
                            << "contains     : " << elements_rep.str() << std::endl
```

Comment (taketwo) [reply common/include/pcl/pcl_tests.h]: Mmm I think yes, this looks more understandable.

**Comment by taketwo** [comment]
Great, looks good to me. Can you please rewrite history to have one commit adding new unit testing macros and second one fixing the bug and improving unit tests? Then we are good to go.

**Comment by Enderdead** [comment]
It's done, tell me if I misunderstood your instructions.

**Pull request merged**